File: leetcode/00200-00299/00206_reverse-linked-list/Solution.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

// Solution 1
// Sample :
//
// Round1
//   head   next
//    ↓      ↓
//    0  ->  1  ->  2  ->  3  ->  4  ->  5
//
func reverseList(head *ListNode) *ListNode {
	var res *ListNode
	for head != nil {
		next := head.Next
		head.Next = res
		res = head
		head = next
	}
	return res
}

func printList(head *ListNode) {
	for head != nil {
		fmt.Println(head.Val)
		head = head.Next
	}
}

func main() {
	l1 := &ListNode{0, nil}
	cur := l1
	for i := 1; i < 10; i++ {
		cur.Next = &ListNode{i, nil}
		cur = cur.Next
	}
	printList(l1)
	fmt.Println("=================")
	res := reverseList(l1)
	printList(res)
}

File: leetcode/00200-00299/00206_reverse-linked-list/Solution.ts
/**
 * ID:    00206
 * Title: Reverse Linked List
 * Difficulty: Easy
 * Description: Given the head of a singly linked list, reverse the list, and return the reversed list.
 *
 * Example 1:
 *
 * Input: head = [1,2,3,4,5] Output: [5,4,3,2,1]
 *
 * Example 2:
 *
 * Input: head = [1,2] Output: [2,1]
 *
 * Example 3:
 *
 * Input: head = [] Output: []
 *
 * Constraints:
 *
 * The number of nodes in the list is the range [0, 5000].
 * -5000 <= Node.val <= 5000
 *
 * Follow up: A linked list can be reversed either iteratively or recursively. Could you implement both?
 */
import { ListNode } from '../../utils/ListNode';
function reverseList(head: ListNode | null): ListNode | null {
  let res: ListNode | null = null
  while (head !== null) {
    let next = head.next
    head.next = res
    res = head
    head = next
  }
  return res
}

function test_00206() {
  let head = ListNode.create([1, 2, 3, 4, 5])
  ListNode.print(reverseList(head))
}

test_00206()
